### PR TITLE
Fix update_release_notes to allow pinning comment to top

### PR DESCRIPTION
**Pull request by AliSoftware**
# Why?

In most of our iOS and Android app repos, we now added a comment at the very top of our `RELEASE-NOTES.txt` file to document to developers how the file should be used / formatted (especially, some instructions about using priority indicators).

The `update_release_notes()` fastlane action, which is used during `code_freeze` lane on all our repos, currently adds the empty section for the next version… at the very top of the `RELEASE-NOTES.txt` file, ignoring that we want this note to stay pinned on top. I got fed up of fixing this manually and moving our instructions comments back to the top myself, so here goes the fix.

# How?

This PR now makes the `update_release_notes` action skip any lines at the top of the `RELEASE-NOTES.txt` file if they start with either `***` (the marker we currently start our comment with in the release note files of most of our repos) or `//` (by convention, as it's the more common marker for comments in many languages). It will then add the new empty section only _after_ those "pinned-to-top comments" lines.

# Testing

1. In one of our iOS app's repo (WPiOS, WCiOS, …), create a new dummy git branch, and `git push -u origin <branchname>` it to the remote.
2. Point the app repo's `fastlane/Pluginfile` to this branch of the Release Toolkit and run `bundle install`.
3. _[Optional]_ Add more comment lines at the top of your `RELEASE-NOTES.txt` – some starting with `***`, some starting with `//`, and with some empty lines too, to check this will still work if we add more pinned comments in the future — and commit them.
4. Run `bundle exec fastlane run ios_update_release_notes new_version:123.4`.
   * The action will add the new entry **for `123.5`** – i.e. the version after the `123.4` that we're simulating a code freeze for.
   * It will then commit and push your branch to the remote – this is why we did `push -u` on step 1, to avoid this last step to crash on the push attempt if it was not tracking a remote branch already.
5. Run `git show HEAD` and check that the new entry for `123.5` in `RELEASE-NOTES.txt` has been added **_not_ at the very top, but _below_ the topmost instructions** in that file (and directly above the section title for the latest version entry you add in those release notes).
6. Delete the git branch you used for those tests, **both locally and on the remote**.
7. Repeat for an Android app's repo, this time running the `bundle exec fastlane run android_update_release_notes new_version:123.4` action instead of the ios one.
8. Repeat the tests on a `RELEASE-NOTES.txt` file which doesn't have a "header" / "comment-on-top" at all (i.e. first line is immediately a version number, and second line is `-----`).

#### Example

<details><summary>State of my modified release notes (as per optional step 3 above) before running the action</summary>

```
$ head -n 20 RELEASE-NOTES.txt
// This file contains the release notes for WooCommerce iOS

// Please add a new bullet point under the corresponding version every time you
// do a new PR that is worth adding a release note entry for.
//
// Thanks.

*** PLEASE FOLLOW THIS FORMAT: [<priority indicator, more stars = higher priority>] <description> [<PR URL>]

5.9
-----
- [*] Removed fulfillment screen and moved fulfillment to the order details screen. [https://github.com/woocommerce/woocommerce-ios/pull/3453]
- [*] Fix: billing information action sheets now are presented correctly on iPad. [https://github.com/woocommerce/woocommerce-ios/pull/3457]
- [*] fix: the rows in the product search list now don't have double separators. [https://github.com/woocommerce/woocommerce-ios/pull/3456]
- [*] fix: when adding a note to an order, the text gets no more deleted if you tap on “Email note to customer”. [https://github.com/woocommerce/woocommerce-ios/pull/3473]


5.8
-----
- [***] Products M5 features are now available to all. Products M5 features: add and edit linked products, add and edit downloadable files, product deletion. [https://github.com/woocommerce/woocommerce-ios/pull/3420]

```

</details>

<details><summary>Running the action</summary>

```
$ bundle exec fastlane run ios_update_release_notes new_version:123.4
[…]
[21:38:40]: --------------------------------------
[21:38:40]: --- Step: ios_update_release_notes ---
[21:38:40]: --------------------------------------
[21:38:40]: Updating the release notes...
[21:38:40]: $ git add ./RELEASE-NOTES.txt
[21:38:40]: $ git commit -m Update\ release\ notes
[21:38:40]: ▸ [tooling/release-notes f10ece833] Update release notes
[21:38:40]: ▸ 1 file changed, 11 insertions(+)
[21:38:40]: $ git push origin HEAD
[21:38:45]: ▸ remote:
[21:38:45]: ▸ remote: GitHub found 1 vulnerability on woocommerce/woocommerce-ios's default branch (1 low). To find out more, visit:
[21:38:45]: ▸ remote:      https://github.com/woocommerce/woocommerce-ios/security/dependabot/Gemfile.lock/nokogiri/open
[21:38:45]: ▸ remote:
[21:38:45]: ▸ To github.com:woocommerce/woocommerce-ios.git
[21:38:45]: ▸ e849a682a..f10ece833  HEAD -> tooling/release-notes
[21:38:45]: Done.
[21:38:45]: Result: true
```

</details>

<details><summary>Resulting diff, where new version's entry has properly been added <i>after</i> the header</summary>

```
$ git show HEAD
commit 00f3826f72e4ae258bb954c3a523e39728ed03fb (HEAD -> dummy/release-notes-test, origin/dummy/release-notes-test)
Author: Olivier Halligon <olivier.halligon@automattic.com>
Date:   Thu Jan 14 21:50:26 2021 +0100

    Update release notes

diff --git a/RELEASE-NOTES.txt b/RELEASE-NOTES.txt
index 8f6cbea52..33d5d770d 100644
--- a/RELEASE-NOTES.txt
+++ b/RELEASE-NOTES.txt
@@ -7,6 +7,10 @@
 
 *** PLEASE FOLLOW THIS FORMAT: [<priority indicator, more stars = higher priority>] <description> [<PR URL>]
 
+123.5
+-----
+
+
 5.9
 -----
 - [*] Removed fulfillment screen and moved fulfillment to the order details screen. [https://github.com/woocommerce/woocommerce-ios/pull/3453]
```
</details>

### CODE REVIEW

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_git_helper.rb
Comment (mokagio) [inline lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_git_helper.rb]: Nice that, on top of the new feature, you replaced `Action.sh` with pure Ruby 👍

**Review by mokagio — DISMISSED** [review]
Works as described 👍 

Screen Shot 2021-01-15 at 4 21 59 pm

Refinement suggestion, take it or leave it. The logic that takes the file lines and new version header as input and generates the new release notes content with pinned comments as output could be extracted and unit tested.

**Comment by AliSoftware** [comment]
I like the idea of adding unit tests for this part. Though for now there's no real good existing place to write them (feels strange to put them in the corresponding `git_helper_spec` file as they would be unrelated to anything `git`…) – which is exactly one of the points paaHJt-1JR-p2 plans to improve.
I'll see where I will add those tests to address this before merging, but I'll likely them in a place that will be temporary until The Great Refactor™ coming soon finds a better home for them.

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_git_helper.rb
Comment (AliSoftware) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_git_helper.rb]: Hehe yep! The next phase is to extract the remaining `Action.sh` which do the commit + push stuff, so that git-related code is kept in `*_git_helper.rb` but non-git related code like above is moved in a better place/file. But that's something that The Great Refactor™ (paaHJt-1JR-p2) will take care of in a separate PR.

**Comment by AliSoftware** [comment]
@mokagio I pushed a new commit that changes quite a lot of things in the code, as I started to include part of my refactoring work into this new commit and this PR now, to be able to properly write the nice new specs for it:

* Removed the code related to release notes from `android/android_git_helper.rb` and `ios/ios_git_helper.rb`, and moved it instead to separate it into:
  * The code only related to the release notes update, moved into a new, common `release_notes_helper.rb` file.
  * The code only related to git, moved into the existing, common `git_helper.rb` file (not the platform-specific ones).
  * Then updated the `actions/{android,ios}/{android,ios}_update_release_notes.rb` actions & call sites accordingly.
* Added unit tests / specs for both of those 2 files and their new methods
  * New `release_notes_helper_spec.rb`
  * Enriched `git_helper_spec.rb` with tests for the new `commit(message:files:push:)` method
  * Moved some helper methods that were created for `ios_lint_localizations_spec` to DRY them into `spec_helper`

Given this, I'll discard your previous approval and let you re-review and re-test the new code.

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_update_release_notes.rb
Comment (mokagio) [inline lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_update_release_notes.rb]: Not necessarily in scope for this change set, but, what do you think of

```suggestion
          Fastlane::Helper::GitHelper.commit(message: "Update release notes with #{next_version} section", files: path, push: true)
```

or

```suggestion
          Fastlane::Helper::GitHelper.commit(message: "Add empty #{next_version} section to release notes", files: path, push: true)
```

Or something else that adds the version number in the commit message for extra context.

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_update_release_notes.rb
Comment (mokagio) [inline lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_update_release_notes.rb]: Same suggestion as the Android counter-part: it might be useful to add the version to the commit title.

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/release_notes_helper.rb
Comment (mokagio) [inline lib/fastlane/plugin/wpmreleasetoolkit/helper/release_notes_helper.rb]:
```suggestion
      # @param [String] path The path to the release notes text file.
```

**Review by mokagio — APPROVED** [review]
Everything looks much better now ✨ 

I repeated the tests on my test branches on Simplenote iOS and Android `release-toolkit-187-test` and they worked as expected.

I noticed an edge case: running the action twice will result in adding the version section twice. I don't think this would happen IRL, because we have pre-checks that would not put us in the state where we're creating a release that already exists. The code is also written in a generic way - `add_new_section` makes no claim of keeping the release notes file "consistent" - so I don't consider this an issue.

---

I won't merge this because I'd like to hear your thoughts on my commit message comment. If you like it, I won't be disappointed if you were to decide to merge this first and address that later; it's really a minor detail.

**Comment by AliSoftware** [comment]
> I noticed an edge case: running the action twice will result in adding the version section twice.

True… but that was already the behavior before 😛 so I don't think it's worth adding any protection about this. Especially given, as you said, that it should never happen upstream in real life as our prechecks preventing us to be in such state.

_Side note: one thing that might be added in `release_notes_helper` in the future (during The Great Refactor™) _might_ be methods like `extract_notes(version:…)` or `extract_latest_version_notes()` or something which might be useful for things like creating the draft notes for the first beta to ADC/GPC (the very first beta we do before the release notes have time to go thru Editorial for review), so if and when I'll be tackling such feature in a future PR as part of this project I might have a method around to extract the title of the topmost section in that file and use it to add checks there. But that's for a future PR and not worth including there yet imho._

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_update_release_notes.rb
Comment (AliSoftware) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_update_release_notes.rb]: Good idea! Done in 310908ddad862919ded5418a493226ccc4bffd41

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_update_release_notes.rb
Comment (AliSoftware) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_update_release_notes.rb]: Done in 310908ddad862919ded5418a493226ccc4bffd41 !

**Pull request merged**
